File: 725.split-linked-list-in-parts.py
#
# @lc app=leetcode id=725 lang=python3
#
# [725] Split Linked List in Parts
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import List, Optional


class Solution:
    '''
    First find the length of the linked list
    Then each part should have n//k nodes
    Except the front n%k ones will have 1 more    
    '''
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        temp = head
        n = 0
        while temp:
            n += 1
            temp = temp.next
        result = []
        temp = head
        for j in range(k):
            temp_head = temp
            if j < n % k:
                thres = n // k + 1
            else:
                thres = n // k
            if thres > 0:
                count = 0
                while count < thres:
                    count += 1
                    if count == thres:
                        temp_tail = temp
                    temp = temp.next
                temp_tail.next = None
            result.append(temp_head)
        return result
# @lc code=end

File: 725.split-linked-list-in-parts.cs
/*
 * @lc app=leetcode id=725 lang=csharp
 *
 * [725] Split Linked List in Parts
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val = 0, ListNode next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode[] SplitListToParts(ListNode head, int k) {
        ListNode temp = head;
        int n = 0;
        
        while (temp != null) {
            n++;
            temp = temp.next;
        }
        
        ListNode[] result = new ListNode[k];
        temp = head;
        
        for (int j = 0; j < k; j++) {
            ListNode tempHead = temp;
            int thres = (j < n % k) ? n / k + 1 : n / k;
            
            if (thres > 0) {
                int count = 0;
                ListNode tempTail = null;
                
                while (count < thres) {
                    count++;
                    
                    if (count == thres) {
                        tempTail = temp;
                    }
                    
                    temp = temp.next;
                }
                
                if (tempTail != null) {
                    tempTail.next = null;
                }
                
            }
            
            result[j] = tempHead;
        }
        
        return result;
    }
}

// @lc code=end

